Tile issue

Hello, while using the quicksettings tile I noticed that it didn't seem to work, here is the error in the logs.

```
2024-06-01 09:15:36.653  4242-4293  AndroidRuntime          xyz.malkki.neostumbler.fdroid        E  FATAL EXCEPTION: DefaultDispatcher-worker-5 (Ask Gemini)
   Process: xyz.malkki.neostumbler.fdroid, PID: 4242
   java.lang.SecurityException: Starting FGS with type location callerApp=ProcessRecord{3012e3a 4242:xyz.malkki.neostumbler.fdroid/u0a396} targetSDK=34 requires permissions: all of the permissions allOf=true [android.permission.FOREGROUND_SERVICE_LOCATION] any of the permissions allOf=false [android.permission.ACCESS_COARSE_LOCATION, android.permission.ACCESS_FINE_LOCATION]  and the app must be in the eligible state/exemptions to access the foreground only permission
   	at android.os.Parcel.createExceptionOrNull(Parcel.java:3242)
   	at android.os.Parcel.createException(Parcel.java:3226)
   	at android.os.Parcel.readException(Parcel.java:3209)
   	at android.os.Parcel.readException(Parcel.java:3151)
   	at android.app.IActivityManager$Stub$Proxy.setServiceForeground(IActivityManager.java:7164)
   	at android.app.Service.startForeground(Service.java:776)
   	at t4.w.q(Unknown Source:365)
   	at a3.a.p(Unknown Source:7)
   	at s3.L.run(Unknown Source:114)
   	at z3.a.run(Unknown Source:95)
   	Suppressed: x3.g: [t0{Cancelling}@4db001e, Dispatchers.Default]
   Caused by: android.os.RemoteException: Remote stack trace:
   	at com.android.server.am.ActiveServices.validateForegroundServiceType(ActiveServices.java:2882)
   	at com.android.server.am.ActiveServices.setServiceForegroundInnerLocked(ActiveServices.java:2569)
   	at com.android.server.am.ActiveServices.setServiceForegroundLocked(ActiveServices.java:1835)
   	at com.android.server.am.ActivityManagerService.setServiceForeground(ActivityManagerService.java:14077)
   	at android.app.IActivityManager$Stub.onTransact(IActivityManager.java:3578)
```

All permissions have been granted.

--- a/app/src/full/AndroidManifest.xml
+++ b/app/src/full/AndroidManifest.xml
@@ -2,7 +2,6 @@
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
     xmlns:tools="http://schemas.android.com/tools">
 
-    <uses-permission android:name="android.permission.ACCESS_BACKGROUND_LOCATION" tools:node="merge" />
     <uses-permission android:name="android.permission.ACTIVITY_RECOGNITION" tools:node="merge"/>
 
     <application tools:node="merge">

--- a/app/src/main/AndroidManifest.xml
+++ b/app/src/main/AndroidManifest.xml
@@ -4,17 +4,26 @@
 
     <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION" />
     <uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION" />
+    <uses-permission android:name="android.permission.ACCESS_BACKGROUND_LOCATION" />
+
     <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
     <uses-permission android:name="android.permission.ACCESS_WIFI_STATE" />
     <uses-permission android:name="android.permission.CHANGE_WIFI_STATE" />
+
     <uses-permission android:name="android.permission.INTERNET" />
+
     <uses-permission android:name="android.permission.FOREGROUND_SERVICE" />
+    <uses-permission android:name="android.permission.FOREGROUND_SERVICE_LOCATION" />
     <uses-permission android:name="android.permission.FOREGROUND_SERVICE_DATA_SYNC" />
+
     <uses-permission android:name="android.permission.POST_NOTIFICATIONS" />
+
     <uses-permission android:name="android.permission.READ_PHONE_STATE" />
+
     <uses-permission android:name="android.permission.REQUEST_IGNORE_BATTERY_OPTIMIZATIONS" />
     <uses-permission android:name="android.permission.RECEIVE_BOOT_COMPLETED" />
     <uses-permission android:name="android.permission.WAKE_LOCK" />
+
     <uses-permission android:name="android.permission.BLUETOOTH_SCAN"/>
     <uses-permission android:name="android.permission.BLUETOOTH" android:maxSdkVersion="30" />
     <uses-permission android:name="android.permission.BLUETOOTH_ADMIN" android:maxSdkVersion="30" />

--- a/app/src/main/java/xyz/malkki/neostumbler/MainActivity.kt
+++ b/app/src/main/java/xyz/malkki/neostumbler/MainActivity.kt
@@ -33,6 +33,11 @@ import xyz.malkki.neostumbler.ui.screens.settings.SettingsScreen
 import xyz.malkki.neostumbler.ui.theme.NeoStumblerTheme
 
 class MainActivity : ComponentActivity() {
+    companion object {
+        const val EXTRA_START_SCANNING = "start_scanning"
+        const val EXTRA_REQUEST_BACKGROUND_PERMISSION = "request_background_permission"
+    }
+
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
 

--- a/app/src/main/java/xyz/malkki/neostumbler/scanner/ScannerService.kt
+++ b/app/src/main/java/xyz/malkki/neostumbler/scanner/ScannerService.kt
@@ -8,6 +8,7 @@ import android.app.PendingIntent
 import android.app.Service
 import android.content.Context
 import android.content.Intent
+import android.content.pm.ServiceInfo
 import android.net.wifi.ScanResult
 import android.net.wifi.WifiManager
 import android.net.wifi.WifiManager.WifiLock
@@ -324,7 +325,7 @@ class ScannerService : Service() {
             }
         }
 
-        startForeground(NOTIFICATION_ID, createNotification())
+        startForeground(NOTIFICATION_ID, createNotification(), ServiceInfo.FOREGROUND_SERVICE_TYPE_LOCATION)
     }
 
     private fun stopScanning() {

--- a/app/src/main/java/xyz/malkki/neostumbler/scanner/quicksettings/ScannerTileService.kt
+++ b/app/src/main/java/xyz/malkki/neostumbler/scanner/quicksettings/ScannerTileService.kt
@@ -1,5 +1,6 @@
 package xyz.malkki.neostumbler.scanner.quicksettings
 
+import android.Manifest
 import android.app.PendingIntent
 import android.content.ComponentName
 import android.content.Intent
@@ -8,10 +9,13 @@ import android.os.Build
 import android.os.IBinder
 import android.service.quicksettings.Tile
 import android.service.quicksettings.TileService
+import androidx.core.service.quicksettings.PendingIntentActivityWrapper
+import androidx.core.service.quicksettings.TileServiceCompat
 import kotlinx.coroutines.runBlocking
 import timber.log.Timber
 import xyz.malkki.neostumbler.MainActivity
 import xyz.malkki.neostumbler.StumblerApplication
+import xyz.malkki.neostumbler.extensions.checkMissingPermissions
 import xyz.malkki.neostumbler.scanner.ScannerService
 import xyz.malkki.neostumbler.utils.OneTimeActionHelper
 import xyz.malkki.neostumbler.utils.PermissionHelper
@@ -84,29 +88,33 @@ class ScannerTileService : TileService() {
 
     override fun onClick() {
         if (!scanningActive) {
-            if (PermissionHelper.hasScanPermissions(this)) {
+            if (PermissionHelper.hasScanPermissions(this)
+                && (Build.VERSION.SDK_INT < Build.VERSION_CODES.UPSIDE_DOWN_CAKE || checkMissingPermissions(Manifest.permission.ACCESS_BACKGROUND_LOCATION).isEmpty())) {
                 //If we already have required permissions, start scanning
                 startForegroundService(ScannerService.startIntent(this))
             } else {
-                //Otherwise open main activity where user can start scanning manually
-                startMainActivity()
+                //Otherwise open main activity to request permissions before starting scanning
+                startMainActivity(requestBackgroundPermission = backgroundLocationPermissionNeeded())
             }
         } else {
             startService(ScannerService.stopIntent(this))
         }
     }
 
-    private fun startMainActivity() {
+    private fun backgroundLocationPermissionNeeded(): Boolean {
+        //As of Android 14, background location permission is needed to start foreground services using location from quick settings tiles
+        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE
+                && checkMissingPermissions(Manifest.permission.ACCESS_BACKGROUND_LOCATION).isNotEmpty()
+    }
+
+    private fun startMainActivity(requestBackgroundPermission: Boolean = false) {
         val intent = Intent(this, MainActivity::class.java).apply {
             flags = Intent.FLAG_ACTIVITY_NEW_TASK
+            putExtra(MainActivity.EXTRA_START_SCANNING, true)
+            putExtra(MainActivity.EXTRA_REQUEST_BACKGROUND_PERMISSION, requestBackgroundPermission)
         }
+        val intentWrapper = PendingIntentActivityWrapper(this, MAIN_ACTIVITY_REQUEST_CODE, intent, PendingIntent.FLAG_UPDATE_CURRENT, false)
 
-        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
-            val pendingIntent = PendingIntent.getActivity(this, MAIN_ACTIVITY_REQUEST_CODE, intent, PendingIntent.FLAG_IMMUTABLE or PendingIntent.FLAG_UPDATE_CURRENT)
-
-            startActivityAndCollapse(pendingIntent)
-        } else {
-            startActivityAndCollapse(intent)
-        }
+        TileServiceCompat.startActivityAndCollapse(this, intentWrapper)
     }
 }

--- a/app/src/main/java/xyz/malkki/neostumbler/ui/screens/ReportsScreen.kt
+++ b/app/src/main/java/xyz/malkki/neostumbler/ui/screens/ReportsScreen.kt
@@ -24,6 +24,7 @@ import androidx.compose.material3.Button
 import androidx.compose.material3.Icon
 import androidx.compose.material3.Text
 import androidx.compose.runtime.Composable
+import androidx.compose.runtime.LaunchedEffect
 import androidx.compose.runtime.livedata.observeAsState
 import androidx.compose.runtime.mutableStateOf
 import androidx.compose.runtime.remember
@@ -41,11 +42,13 @@ import androidx.compose.ui.unit.dp
 import androidx.compose.ui.unit.sp
 import androidx.lifecycle.viewmodel.compose.viewModel
 import kotlinx.coroutines.launch
+import xyz.malkki.neostumbler.MainActivity
 import xyz.malkki.neostumbler.R
 import xyz.malkki.neostumbler.StumblerApplication
 import xyz.malkki.neostumbler.db.entities.ReportWithStats
 import xyz.malkki.neostumbler.extensions.checkMissingPermissions
 import xyz.malkki.neostumbler.extensions.defaultLocale
+import xyz.malkki.neostumbler.extensions.getActivity
 import xyz.malkki.neostumbler.scanner.ScannerService
 import xyz.malkki.neostumbler.scanner.quicksettings.ScannerTileService
 import xyz.malkki.neostumbler.ui.composables.AddQSTileDialog
@@ -122,6 +125,7 @@ private val requiredPermissions = mutableListOf<String>()
 @Composable
 fun ForegroundScanningButton() {
     val context = LocalContext.current
+    val intent = context.getActivity()?.intent
 
     val coroutineScope = rememberCoroutineScope()
     val oneTimeActionHelper = OneTimeActionHelper(context.applicationContext as StumblerApplication)
@@ -136,6 +140,10 @@ fun ForegroundScanningButton() {
         mutableStateOf(false)
     }
 
+    val showBackgroundLocationPermissionDialog = remember {
+        mutableStateOf(false)
+    }
+
     val showQuickSettingsDialog = remember {
         mutableStateOf(false)
     }
@@ -144,11 +152,42 @@ fun ForegroundScanningButton() {
         context.checkMissingPermissions(*requiredPermissions)
     }
 
+    fun startScanning() {
+        if (ScannerService.serviceRunning) {
+            context.startService(ScannerService.stopIntent(context))
+
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
+                //Prompt user to add the quick settings tile for scanning
+                coroutineScope.launch {
+                    if (!oneTimeActionHelper.hasActionBeenShown(ScannerTileService.ADD_QS_TILE_ACTION_NAME)) {
+                        showQuickSettingsDialog.value = true
+                    }
+                }
+            }
+        } else {
+            if (Manifest.permission.ACCESS_FINE_LOCATION !in missingPermissions) {
+                if (intent?.getBooleanExtra(MainActivity.EXTRA_REQUEST_BACKGROUND_PERMISSION, false) == true
+                    && context.checkMissingPermissions(Manifest.permission.ACCESS_BACKGROUND_LOCATION).isNotEmpty()) {
+                    showBackgroundLocationPermissionDialog.value = true
+                } else {
+                    showBatteryOptimizationsDialog.value = true
+                }
+            } else {
+                showPermissionDialog.value = true
+            }
+        }
+    }
+
     val onPermissionsGranted: (Map<String, Boolean>) -> Unit = { permissions ->
         showPermissionDialog.value = false
 
         if (permissions[Manifest.permission.ACCESS_FINE_LOCATION] == true) {
-            showBatteryOptimizationsDialog.value = true
+            if (intent?.getBooleanExtra(MainActivity.EXTRA_REQUEST_BACKGROUND_PERMISSION, false) == true
+                && context.checkMissingPermissions(Manifest.permission.ACCESS_BACKGROUND_LOCATION).isNotEmpty()) {
+                showBackgroundLocationPermissionDialog.value = true
+            } else {
+                showBatteryOptimizationsDialog.value = true
+            }
         } else {
             Toast.makeText(context, context.getString(R.string.permissions_not_granted), Toast.LENGTH_SHORT).show()
         }
@@ -176,6 +215,19 @@ fun ForegroundScanningButton() {
         )
     }
 
+    if (showBackgroundLocationPermissionDialog.value) {
+        PermissionsDialog(
+            missingPermissions = listOf(Manifest.permission.ACCESS_BACKGROUND_LOCATION),
+            permissionRationales = mapOf(
+                Manifest.permission.ACCESS_BACKGROUND_LOCATION to stringResource(id = R.string.permission_rationale_background_location_quick_settings)
+            ),
+            onPermissionsGranted = {
+                showBackgroundLocationPermissionDialog.value = false
+
+                showBatteryOptimizationsDialog.value = true
+            })
+    }
+
     if (showBatteryOptimizationsDialog.value) {
         BatteryOptimizationsDialog(onBatteryOptimizationsDisabled = {
             showBatteryOptimizationsDialog.value = false
@@ -183,26 +235,15 @@ fun ForegroundScanningButton() {
         })
     }
 
+    LaunchedEffect(intent) {
+        if (intent?.getBooleanExtra(MainActivity.EXTRA_START_SCANNING, false) == true) {
+            startScanning()
+        }
+    }
+
     Button(
         onClick = {
-            if (ScannerService.serviceRunning) {
-                context.startService(ScannerService.stopIntent(context))
-
-                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
-                    //Prompt user to add the quick settings tile for scanning
-                    coroutineScope.launch {
-                        if (!oneTimeActionHelper.hasActionBeenShown(ScannerTileService.ADD_QS_TILE_ACTION_NAME)) {
-                            showQuickSettingsDialog.value = true
-                        }
-                    }
-                }
-            } else {
-                if (Manifest.permission.ACCESS_FINE_LOCATION !in missingPermissions) {
-                    showBatteryOptimizationsDialog.value = true
-                } else {
-                    showPermissionDialog.value = true
-                }
-            }
+            startScanning()
         }
     ) {
         val stringResId = if (serviceConnection.value != null) {

--- a/app/src/main/res/values/strings.xml
+++ b/app/src/main/res/values/strings.xml
@@ -75,4 +75,8 @@
 
     <string name="endpoint">Endpoint</string>
     <string name="api_key">API key (optional)</string>
+
+    <string name="permission_rationale_background_location_quick_settings">
+        Starting scanning from the quick settings needs access to location in the background
+    </string>
 </resources>

